File: other/boto3.mjs
/* 
V2: https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/getting-started-nodejs.html
V3: https://aws.amazon.com/es/sdk-for-javascript/

https://huggingface.co/openai/whisper-tiny
https://huggingface.co/models?pipeline_tag=automatic-speech-recognition&language=es,en&sort=trending
https://huggingface.co/docs/transformers.js/api/pipelines#module_pipelines.AutomaticSpeechRecognitionPipeline
https://huggingface.co/docs/transformers.js/guides/node-audio-processing
*/


import {
  ListObjectsCommand,
  GetObjectCommand,
  S3Client,
} from "@aws-sdk/client-s3";
import { Readable } from "stream";
import fs from "fs";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";

import { pipeline } from "@xenova/transformers";
import wavefile from "wavefile";
import fetch from "node-fetch";

const client = new S3Client({
  region: "us-east-1",
});

const bucketName = "tnknbucket";
const command = new ListObjectsCommand({ Bucket: bucketName });

const getObject = async () => {
  try {
    // const { Body } = await client.send(
    //   new GetObjectCommand({
    //     Bucket: bucketName,
    //     Key: "audio/recorded_audio.wav",
    //   })
    // );
    // console.log(Body);

    // get signed URL of the object
    const command = new GetObjectCommand({
      Bucket: bucketName,
      Key: "audio/recorded_audio.wav",
    });

    // No funciona porque el audio grabado desde el navegador tiene un formato diferente
    let url = await getSignedUrl(client, command, { expiresIn: 3600 });
    console.log(url);

    url = "https://huggingface.co/datasets/Xenova/transformers.js-docs/resolve/main/jfk.wav";
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`unexpected response ${response.statusText}`);
    }
    // console.log(response);
    let buffer = Buffer.from(await fetch(url).then((x) => x.arrayBuffer()));
    console.log(buffer);

    let wav = new wavefile.WaveFile(buffer);
    wav.toBitDepth("32f"); // Pipeline expects input as a Float32Array
    wav.toSampleRate(16000); // Whisper expects audio with a sampling rate of 16000
    let audioData = wav.getSamples();
    if (Array.isArray(audioData)) {
      if (audioData.length > 1) {
        const SCALING_FACTOR = Math.sqrt(2);

        // Merge channels (into first channel to save memory)
        for (let i = 0; i < audioData[0].length; ++i) {
          audioData[0][i] =
            (SCALING_FACTOR * (audioData[0][i] + audioData[1][i])) / 2;
        }
      }

      // Select first channel
      audioData = audioData[0];
    }

    const transcriber = await pipeline(
      "automatic-speech-recognition",
      "Xenova/whisper-tiny.en"
    );

    let start = performance.now();
    let output = await transcriber(audioData, {
      // language: "spanish",
      return_timestamps: true,
      // task: "transcribe",
    });
    let end = performance.now();
    console.log(`Execution duration: ${(end - start) / 1000} seconds`);

    console.log(output);

    // // Read the stream and convert it to bytes
    // // If uncoomment the next 2 lines before, the stream will be consumed and the next code will fail
    // const data = await streamToBuffer(Body);
    // console.log(data);

    // // Create a writable stream to the file
    // const writableStream = fs.createWriteStream("recorded_audio.wav");

    // // Pipe the S3 stream to the file stream
    // Body.pipe(writableStream);

    // // Wait for the write operation to finish
    // await new Promise((resolve, reject) => {
    //   writableStream.on("finish", resolve);
    //   writableStream.on("error", reject);
    // });

    // console.log("WAV file saved successfully.");
  } catch (error) {
    console.error(error);
  }
};

getObject();

File: other/boto3.py
# https://replicate.com/vaibhavs10/incredibly-fast-whisper?prediction=4hbuid3bzbzuarnkxnr6anwyi4&output=json&input=python
# https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3-presigned-urls.html
# https://docs.aws.amazon.com/code-library/latest/ug/python_3_code_examples.html

"""
aws configure
aws sts get-caller-identity
aws s3 ls
"""

import json
import logging
from botocore.exceptions import ClientError
import boto3
import replicate
from dotenv import load_dotenv
from datetime import datetime
load_dotenv()


def custom_serializer(obj):
    """Serialize datetime objects to string."""
    if isinstance(obj, datetime):
        return obj.isoformat()


def create_presigned_url(bucket_name, object_name, expiration=3600):
    """Generate a presigned URL to share an S3 object

    :param bucket_name: string
    :param object_name: string
    :param expiration: Time in seconds for the presigned URL to remain valid
    :return: Presigned URL as string. If error, returns None.
    """

    # Generate a presigned URL for the S3 object
    s3_client = boto3.client('s3')
    try:
        response = s3_client.generate_presigned_url('get_object',
                                                    Params={'Bucket': bucket_name,
                                                            'Key': object_name},
                                                    ExpiresIn=expiration)
    except ClientError as e:
        logging.error(e)
        return None

    # The response contains the presigned URL
    return response

# -----------------------------------------------------------------------------
# AWS S3
# -----------------------------------------------------------------------------


s3 = boto3.client('s3')
response = s3.list_buckets()

print('Existing buckets:')
for bucket in response['Buckets']:
    print(f'- {bucket["Name"]}')

bucket_name = "tnknbucket"
print(json.dumps(s3.list_objects(Bucket=bucket_name),
                 default=custom_serializer,
                 indent=2))

presigned_url = create_presigned_url(bucket_name, "audio/recorded_audio.wav")
print(presigned_url)

# -----------------------------------------------------------------------------
# Replicate Incredibly Fast Whisper: Transcribe from presigned URL
# -----------------------------------------------------------------------------

output = replicate.run(
    "vaibhavs10/incredibly-fast-whisper:3ab86df6c8f54c11309d4d1f930ac292bad43ace52d10c80d87eb258b3c9f79c",
    input={
        "task": "transcribe",
        "audio": presigned_url,
        "language": "None",
        "timestamp": "chunk",
        "batch_size": 64,
        "diarise_audio": False
    }
)

print(json.dumps(output, indent=2, ensure_ascii=False))
